libs/filamentjs/jsbindings.cpp

```diff
@@ -0,0 +1,373 @@
+        .value("UV1", UV1)
+        .value("BONE_INDICES", BONE_INDICES)
+        .value("BONE_WEIGHTS", BONE_WEIGHTS);
+
+ enum_<VertexBuffer::AttributeType>("VertexBuffer$AttributeType")
+        .value("BYTE", VertexBuffer::AttributeType::BYTE)
+        .value("BYTE2", VertexBuffer::AttributeType::BYTE2)
+        .value("BYTE3", VertexBuffer::AttributeType::BYTE3)
+        .value("BYTE4", VertexBuffer::AttributeType::BYTE4)
+        .value("UBYTE", VertexBuffer::AttributeType::UBYTE)
+        .value("UBYTE2", VertexBuffer::AttributeType::UBYTE2)
+        .value("UBYTE3", VertexBuffer::AttributeType::UBYTE3)
+        .value("UBYTE4", VertexBuffer::AttributeType::UBYTE4)
+        .value("SHORT", VertexBuffer::AttributeType::SHORT)
+        .value("SHORT2", VertexBuffer::AttributeType::SHORT2)
+        .value("SHORT3", VertexBuffer::AttributeType::SHORT3)
+        .value("SHORT4", VertexBuffer::AttributeType::SHORT4)
+        .value("USHORT", VertexBuffer::AttributeType::USHORT)
+        .value("USHORT2", VertexBuffer::AttributeType::USHORT2)
+        .value("USHORT3", VertexBuffer::AttributeType::USHORT3)
+        .value("USHORT4", VertexBuffer::AttributeType::USHORT4)
+        .value("INT", VertexBuffer::AttributeType::INT)
+        .value("UINT", VertexBuffer::AttributeType::UINT)
+        .value("FLOAT", VertexBuffer::AttributeType::FLOAT)
+        .value("FLOAT2", VertexBuffer::AttributeType::FLOAT2)
+        .value("FLOAT3", VertexBuffer::AttributeType::FLOAT3)
+        .value("FLOAT4", VertexBuffer::AttributeType::FLOAT4)
+        .value("HALF", VertexBuffer::AttributeType::HALF)
+        .value("HALF2", VertexBuffer::AttributeType::HALF2)
+        .value("HALF3", VertexBuffer::AttributeType::HALF3)
+        .value("HALF4", VertexBuffer::AttributeType::HALF4);
+
+ enum_<IndexBuffer::IndexType>("IndexBuffer$IndexType")
+        .value("USHORT", IndexBuffer::IndexType::USHORT)
+        .value("UINT", IndexBuffer::IndexType::UINT);
+
+ enum_<RenderableManager::PrimitiveType>("RenderableManager$PrimitiveType")
+        .value("POINTS", RenderableManager::PrimitiveType::POINTS)
+        .value("LINES", RenderableManager::PrimitiveType::LINES)
+        .value("TRIANGLES", RenderableManager::PrimitiveType::TRIANGLES)
+        .value("NONE", RenderableManager::PrimitiveType::NONE);
+
+ enum_<View::DepthPrepass>("View$DepthPrepass")
+        .value("DEFAULT", View::DepthPrepass::DEFAULT)
+        .value("DISABLED", View::DepthPrepass::DISABLED)
+        .value("ENABLED", View::DepthPrepass::ENABLED);
+
+// CORE FILAMENT CLASSES
+// ---------------------
+
+class_<Engine>("Engine")
+    .class_function("_create", (Engine* (*)()) [] { return Engine::create(); },
+            allow_raw_pointers())
+    .class_function("destroy", (void (*)(Engine*)) []
+            (Engine* engine) { Engine::destroy(&engine); }, allow_raw_pointers())
+    .function("execute", &Engine::execute)
+    .function("createSwapChain", (SwapChain* (*)(Engine*)) []
```

Comment:
In the future, it might be nice to allow users to pass in the `canvas` they want to use here as the `SwapChain`. Is it correct that the current implementation assumes there will be a `<canvas id="filament-canvas" />` in the DOM?

Reply:
Currently, clients pass the DOM canvas object into the Engine constructor, because Filament creates GL objects inside of Engine::init.

In the future, I would like to support multi-canvas rendering by passing the canvas into the SwapChain constructor rather than the Engine constructor.  However that will require some refactoring with Filament's internals.